### PR TITLE
Unit Tests for Example files from E8 to E15

### PR DESCRIPTION
**Pull request by SamuilsRulovs**
Create unit tests for:

- [x] brep-model
- [x] csg-primitive
- [x] beam-standard-case
- [x] column-extruded-solid
- [x] wall-standard-case
- [x] structural-curve-member
- [x] beam-varying-cardinal-points
- [x] reinforcing-assembly
- [x] horizontal-alignment
- [x] terrain-surface

Edit a foundation for unit tests:

- [x] construction-scheduling-task (Does not work correctly) (Issue)
- [x] reinforcing-stirrup (Something wrong with BoundingBox)(Issue)

Edit a foundation for unit tests:

- [x] beam-extruded-solid
- [x] beam-revolved-solid
- [x] wall-elemented-case
- [x] wall-with-opening-and-window
- [x] air-terminal-element
- [x] air-terminal-library-object
- [x] beam-varying-extrusion-paths
- [x] beam-varying-profiles
- [x] slab-openings
- [x] bloss-curve
- [x] linear-placement
- [x] sectioned-solid
- [x] terrain-and-alignment
- [x] vertical-alignment

### CODE REVIEW

File: UnitTests/Schemas/IFC4X1/air-terminal-element/new 1 [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+include(CreateUnitTests)
+
+CreateIfcFileUnitTestForSchema(air-terminal-element IFC4X1)
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/air-terminal-element/new 1]: I don't understand, why this is in file `new 1`?

File: UnitTests/Schemas/IFC4X1/air-terminal-element/src/air-terminal-element.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+	// Arrange
+	buw::Image4b image = renderer->captureImage();
+
+	// Act
+	buw::storeImage(testPath("air-terminal-element.png").string(), image);
+
+	// Assert
+	EXPECT_NO_THROW(buw::loadImage4b(testPath("air-terminal-element.png").string()));
+}
+
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/air-terminal-element/src/air-terminal-element.cpp:92]: Check for existence of:

- IFCEXTRUDEDAREASOLIDTAPERED
- IFCEXTRUDEDAREASOLID
- IFCCIRCLEHOLLOWPROFILEDEF
- IFCRECTANGLEHOLLOWPROFILEDEF
- IFCDERIVEDPROFILEDEF

File: UnitTests/Schemas/IFC4X1/air-terminal-library-object/src/air-terminal-library-object.cpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+	// Arrange
+	buw::Image4b image = renderer->captureImage();
+
+	// Act
+	buw::storeImage(testPath("air-terminal-library-object.png").string(), image);
+
+	// Assert
+	EXPECT_NO_THROW(buw::loadImage4b(testPath("air-terminal-library-object.png").string()));
+}
+
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/air-terminal-library-object/src/air-terminal-library-object.cpp:86]: Existence of:

- IFCEXTRUDEDAREASOLIDTAPERED
- IFCRECTANGLEHOLLOWPROFILEDEF

File: UserInterface/Dialogues/PreferencesDialog.cpp [hunk 4]
```diff
@@ -29,8 +29,9 @@
 
 #include <QTranslator>
 
-OpenInfraPlatform::UserInterface::PreferencesDialog::PreferencesDialog(OpenInfraPlatform::UserInterface::View* view, QWidget *parent /*= nullptr*/) :
-	ui_(new Ui::PreferencesDialog),
+OpenInfraPlatform::UserInterface::PreferencesDialog::PreferencesDialog(OpenInfraPlatform::UserInterface::View* view, QWidget* parent /*= nullptr*/) :
+
+	ui_(new PreferencesDialog(view, parent)),
```
Comment (pjanck) [inline UserInterface/Dialogues/PreferencesDialog.cpp]: What is the reason for these changes?

File: UserInterface/Dialogues/PreferencesDialog.h [hunk 5]
```diff
@@ -102,7 +102,7 @@ namespace OpenInfraPlatform
             bool useTextured = false;
 			
 		private:
-			Ui::PreferencesDialog*	ui_;
+			PreferencesDialog*	    ui_;
```
Comment (pjanck) [inline UserInterface/Dialogues/PreferencesDialog.h]: What is the reason for these changes?

**Review by pjanck — CHANGES_REQUESTED** [review]
All tests are missing the check for essential entities.

File: UserInterface/Dialogues/PreferencesDialog.cpp (hunk 4 above)
Comment (SamuilsRulovs) [reply UserInterface/Dialogues/PreferencesDialog.cpp]: Thank you for a hint. I have tried to find a reason, why OpenInfraPlatform.UI is not working. Now I know why 👍

Comment (pjanck) [reply UserInterface/Dialogues/PreferencesDialog.cpp]: Just revert these changes, as they should not have been there :D

**Review by pjanck — COMMENTED** [review]
great work so far!

File: UnitTests/Schemas/IFC4X1/beam-varying-cardinal-points/src/beam-varying-cardinal-points.cpp [hunk 6]
```diff
@@ -53,15 +51,28 @@ class BeamVaryingCardinalPoints : public VisualTest {
...
+	buw::ReferenceCounted<oip::IfcModel> model = buw::makeReferenceCounted<oip::IfcModel>();
 };
 
 TEST_F(BeamVaryingCardinalPoints, AllEntitiesAreRead) {
-	EXPECT_THAT(express_model->entities.size(), Eq(90));
+	EXPECT_THAT(express_model->entities.size(), Eq(94));
+}
+
+TEST_F(BeamVaryingCardinalPoints, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCBEAMSTANDARDCASE"; });
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-varying-cardinal-points/src/beam-varying-cardinal-points.cpp]: We focus on geometrical entities, like extruded solids, profile defs and similar. Please check for those as well.

File: UnitTests/Schemas/IFC4X1/beam-standard-case/src/beam-standard-case.cpp [hunk 7]
```diff
@@ -53,15 +51,27 @@ class BeamStandardCase : public VisualTest {
...
+	buw::ReferenceCounted<oip::IfcModel> model = buw::makeReferenceCounted<oip::IfcModel>();
 };
 
 TEST_F(BeamStandardCase, AllEntitiesAreRead) {
-	EXPECT_THAT(express_model->entities.size(), Eq(349));
+	EXPECT_THAT(express_model->entities.size(), Eq(366));
+}
+
+TEST_F(BeamStandardCase, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCBEAMSTANDARDCASE"; });
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-standard-case/src/beam-standard-case.cpp]: We focus on geometrical entities, like extruded solids, profile defs and similar. Please check for those as well.

File: UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp [hunk 8]
```diff
@@ -53,17 +51,29 @@ class BeamVaryingExtrusionPaths : public VisualTest {
...
-	buw::ReferenceCounted<oip::IfcGeometryModel> model = buw::makeReferenceCounted<oip::IfcGeometryModel>();
+	buw::ReferenceCounted<oip::IfcModel> model = buw::makeReferenceCounted<oip::IfcModel>();
 };
 
 TEST_F(BeamVaryingExtrusionPaths, AllEntitiesAreRead) {
 	EXPECT_THAT(express_model->entities.size(), Eq(67));
 }
 
+TEST_F(BeamVaryingExtrusionPaths, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCBEAMSTANDARDCASE"; });
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp]: We focus on geometrical entities, like extruded solids, profile defs and similar. Please check for those as well.

File: UnitTests/Schemas/IFC4X1/beam-varying-profiles/src/beam-varying-profiles.cpp [hunk 9]
```diff
@@ -53,17 +51,29 @@ class BeamVaryingProfiles : public VisualTest {
...
+	buw::ReferenceCounted<oip::IfcModel> model = buw::makeReferenceCounted<oip::IfcModel>();
 };
 
 TEST_F(BeamVaryingProfiles, AllEntitiesAreRead) {
-	EXPECT_THAT(express_model->entities.size(), Eq(63\));
+	EXPECT_THAT(express_model->entities.size(), Eq(63));
+}
+
+TEST_F(BeamVaryingProfiles, IFCHasAnEssentialEntity) {
+	auto result = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCBEAMSTANDARDCASE"; });
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-varying-profiles/src/beam-varying-profiles.cpp]: We focus on geometrical entities, like extruded solids, profile defs and similar. Please check for those as well.

**Review by pjanck — CHANGES_REQUESTED** [review]

**Comment by pjanck** [comment]
Please denote as ready for review, when all are done.

File: UnitTests/Schemas/IFC4X1/air-terminal-element/new 1 (hunk 1 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/air-terminal-element/new 1]: Resolved: 86271d8

File: UnitTests/Schemas/IFC4X1/air-terminal-library-object/src/air-terminal-library-object.cpp (hunk 3 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/air-terminal-library-object/src/air-terminal-library-object.cpp:86]: Resolved: 86271d8

File: UnitTests/Schemas/IFC4X1/air-terminal-element/src/air-terminal-element.cpp (hunk 2 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/air-terminal-element/src/air-terminal-element.cpp:92]: Resolved: 86271d8

File: UnitTests/Schemas/IFC4X1/beam-varying-profiles/src/beam-varying-profiles.cpp (hunk 9 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-varying-profiles/src/beam-varying-profiles.cpp]: Resolved: 95071ab

File: UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp (hunk 8 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp]: Resolved: 95071ab

File: UnitTests/Schemas/IFC4X1/beam-standard-case/src/beam-standard-case.cpp (hunk 7 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-standard-case/src/beam-standard-case.cpp]: Resolved: 95071ab

File: UnitTests/Schemas/IFC4X1/beam-varying-cardinal-points/src/beam-varying-cardinal-points.cpp (hunk 6 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-varying-cardinal-points/src/beam-varying-cardinal-points.cpp]: Resolved: 95071ab

File: UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp [hunk 10]
```diff
@@ -77,7 +84,7 @@ TEST_F(BeamVaryingExtrusionPaths, CountEssentialEntities) {
 TEST_F(BeamVaryingExtrusionPaths, ImageIsSaved)
 {
 	// Arrange
-	buw::Image4b image = renderer->captureImage();
+	buw::Image4b auto result2 = std::find_if(express_model->entities.begin(), express_model->entities.end(), [](auto &pair) -> bool { return pair.second->classname() == "IFCEXTRUDEDAREASOLID"; });image = renderer->captureImage();
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp]: this doesn't look right

File: UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp [hunk 11]
```diff
@@ -106,14 +113,14 @@ TEST_F(BrepModel, PlaneSurfaceViews)
...
-	/*
+	
 	buw::storeImage(testPath("brep-model_front.png").string(), image_front);
 	buw::storeImage(testPath("brep-model_top.png").string(), image_top);
 	buw::storeImage(testPath("brep-model_bottom.png").string(), image_bottom);
 	buw::storeImage(testPath("brep-model_left.png").string(), image_left);
 	buw::storeImage(testPath("brep-model_right.png").string(), image_right);
 	buw::storeImage(testPath("brep-model_back.png").string(), image_back);
-	*/
+
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp]: Why is this not commented out?

File: UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp [hunk 12]
```diff
@@ -162,7 +169,7 @@ TEST_F(BrepModel, VertexViews)
 	buw::Image4b image_right_bottom_back = CaptureImage();
 
 	// uncomment following lines to also save the screen shot
-	/*
+
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp]: Similarly, should be commented out?

File: UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp [hunk 13]
```diff
@@ -171,7 +178,7 @@ TEST_F(BrepModel, VertexViews)
 	buw::storeImage(testPath("brep-model_top_right_back.png").string(), image_top_right_back);
 	buw::storeImage(testPath("brep-model_back_left_bottom.png").string(), image_back_left_bottom);
 	buw::storeImage(testPath("brep-model_right_bottom_back.png").string(), image_right_bottom_back);
-	*/
+
```
Comment (pjanck) [inline UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp]: Similarly, should be commented out?

**Review by pjanck — CHANGES_REQUESTED** [review]
Small changes required.

Also, all terrrain examples should check for `IfcTriangulatedIrregularNetwork`.

File: UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp (hunk 13 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp]: Resolved: 3e3f66c

File: UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp (hunk 12 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp]: Resolved: 3e3f66c

File: UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp (hunk 11 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/brep-model/src/brep-model.cpp]: Resolved: 3e3f66c

File: UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp (hunk 10 above)
Comment (SamuilsRulovs) [reply UnitTests/Schemas/IFC4X1/beam-varying-extrusion-paths/src/beam-varying-extrusion-paths.cpp]: Resolved: 3e3f66c

**Review by pjanck — APPROVED** [review]
